{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "make environment using r_spatial.yml in Spatial folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "micromamba create -n r_spatial -f r_spatial.yml -c conda-forge --channel-priority strict\n",
    "\n",
    "R\n",
    "\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "\n",
    "BiocManager::install(\"nnSVG\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#locally\n",
    "mamba activate r_spatial\n",
    "rstudio #will open rstudio locally with mounts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Note reasoning for batches is in comparing_outliers.csv  \n",
    " - #0 manual processing in interactive R session on cluster \n",
    " - #1 to process each sample individually via slurm submission - avoiding as it results in different numbers of rows in count matrix of each sample which prevents merging later\n",
    " - #2 to process in bulk up to or including nnsvg step  \n",
    "\n",
    "**skip ahead to 2**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "source": [
    "0. To manually run the preprocessing pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "# #open interactive node session, in terminal connect to cluster then enter\n",
    "\n",
    "srun --pty --job-name=interactive_r --mem=32G --cpus-per-task=4 --partition=epyc --time=024:00:00 bash \n",
    "#activate r environment with\n",
    "micromambda activate r_env\n",
    "#open R\n",
    "R\n",
    "Sys.setenv(PATH = paste(Sys.getenv(\"PATH\"), \"/home/lythgo02/micromamba/envs/r_env\", sep = \":\"))\n",
    "#then run the commands in 20240204_QC_final_workflow.rmd but you will need to adapt the code for the samples you want to process as logbatch mitochondrial filters and the arbitrary mitochondrial filters batch"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. to run by as individual samples, open ssh terminal and set of individual slurm jobs for each (using a different script for the log batch and the arbitrary batch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#1.0\n",
    "\n",
    "cd /mnt/scratchc/fmlab/lythgo02/visium_data/\n",
    "for SAMPLE in SITSA3 SITSB2 SITSB4 SITSC1 SITSC3 SITSD3 SITSE4 SITSF2 SITSF4 ; \n",
    "do sbatch \n",
    "--job-name=${SAMPLE} \n",
    "--output=logs/${SAMPLE}.out \n",
    "--error=logs/${SAMPLE}.err \n",
    "--time=100:00:00 \n",
    "--mem=32G \n",
    "--cpus-per-task=8 \n",
    "--partition=epyc \n",
    "--wrap=\"Rscript single_sample_process_logmito.r ${SAMPLE}\"; \n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#1.2\n",
    "\n",
    "cd /mnt/scratchc/fmlab/lythgo02/visium_data/\n",
    "\n",
    "for SAMPLE in SITSA1 SITSA2 SITSA4 SITSB1 SITSB3 SITSC2 SITSC4 SITSD1 SITSD2 SITSD4 SITSE2 SITSG2 SITSH2 ; \n",
    " do sbatch \n",
    " --job-name=${SAMPLE} \n",
    " --output=logs/${SAMPLE}.out \n",
    " --error=logs/${SAMPLE}.err \n",
    " --time=048:00:00 \n",
    " --mem=32G \n",
    " --cpus-per-task=8 \n",
    " --partition=epyc \n",
    " --wrap=\"Rscript single_sample_processing_arbitrary.r ${SAMPLE}\"; \n",
    " done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "once they have finished running cd into the respectiv folders and run the following"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#1.3\n",
    "\n",
    "cd logbatch \n",
    "\n",
    "for SAMPLE in post_gettophgvSITSA3 post_gettophgvSITSB2 post_gettophgvSITSB4 post_gettophgvSITSC1 post_gettophgvSITSC3 post_gettophgvSITSD3 post_gettophgvSITSE4 post_gettophgvSITSF2 post_gettophgvSITSF4; \n",
    " do     sbatch --job-name=${SAMPLE}            \n",
    " --output=logs/${SAMPLE}.out            \n",
    " --error=logs/${SAMPLE}.err            \n",
    " --time=0100:00:00            \n",
    " --mem=32G            \n",
    " --cpus-per-task=8            \n",
    " --partition=epyc            \n",
    " --wrap=\"Rscript /mnt/scratchc/fmlab/lythgo02/visium_data/logbatch/nnsvg_logbatch.r ${SAMPLE}\"; \n",
    " done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#1.4\n",
    "\n",
    "cd into arbitrary\n",
    "\n",
    "for SAMPLE in post_gettophgvSITSA1 post_gettophgvSITSA2 post_gettophgvSITSA4 post_gettophgvSITSB1 post_gettophgvSITSB3 post_gettophgvSITSC2 post_gettophgvSITSC4 post_gettophgvSITSD1 post_gettophgvSITSD2 post_gettophgvSITSD4 post_gettophgvSITSE2 post_gettophgvSITSG2 post_gettophgvSITSH2;  \n",
    "do     \n",
    "sbatch \n",
    "--job-name=${SAMPLE}            \n",
    "--output=logs/${SAMPLE}.out            \n",
    "--error=logs/${SAMPLE}.err            \n",
    "--time=0100:00:00            \n",
    "--mem=32G            \n",
    "--cpus-per-task=8            \n",
    "--partition=epyc            \n",
    "--wrap=\"Rscript /mnt/scratchc/fmlab/lythgo02/visium_data/arbitrary/nnsvg_arbitrary.r ${SAMPLE}\";\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#1.5 then pick up after running nnsvg \n",
    "#open interactive r session within micromamba r_rnv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#1.5\n",
    "\n",
    "knitr::opts_chunk$set(echo = TRUE)\n",
    "library(spatialLIBD)\n",
    "library(nnSVG)\n",
    "library(SpatialExperiment)  \n",
    "library(scater)\n",
    "library(AnnotationHub)\n",
    "library(tidyverse)\n",
    "library(ggspavis)\n",
    "library(scran)\n",
    "library(DT)\n",
    "projDir = \"/mnt/scratchc/fmlab/lythgo02/visium_data/\"\n",
    "\n",
    "\n",
    "load_all_data <- \"/mnt/scratchc/fmlab/lythgo02/visium_data/logbatch/nnsvg\"\n",
    "# List files in the directory\n",
    "files <- list.files(path = load_all_data, pattern = \"^nnsvg\", full.names = TRUE)\n",
    "\n",
    "# Print out the list of files\n",
    "print(files)\n",
    "\n",
    "#assign names so they are loaded into the spe with the correct identifiers\n",
    "logbatch <- c(\"SITSA3\", \"SITSB2\", \"SITSB4\", \n",
    "                 \"SITSC1\",\"SITSC3\",\"SITSD3\",\n",
    "                 \"SITSE4\",\"SITSF2\", \"SITSF4\")\n",
    "\n",
    "names(files) <- logbatch\n",
    "# Initialize a list to store each sample\n",
    "spe_list <- list()\n",
    "\n",
    "# Loop through each file and process it\n",
    "for (sample_name in names(files)) {\n",
    "  # Read the Visium data\n",
    "  spe <- readRDS(files[sample_name])\n",
    "    # Add unique barcodes\n",
    "  spe$barcode <- rownames(colData(spe))\n",
    "  spe$barcodeid <- gsub(\"-1$\", paste0(\"-\", sample_name), spe$barcode)\n",
    "  rownames(colData(spe)) <- spe$barcodeid\n",
    "  spe$sample_id <- sample_name\n",
    "\n",
    "  colData(spe)$qc_mito <- colData(spe)$log_is_outlier_mad\n",
    "  colData(spe) <- colData(spe)[ , !(colnames(colData(spe)) %in% c(\"log_mito_percent\",\"log_median_mito_percent\",\"log_mad_mito_percent\",\"log_deviation_above_median\",\"log_is_outlier_mad\"))]\n",
    "\n",
    "  # Store the processed sample in the list\n",
    "  spe_list[[sample_name]] <- spe \n",
    "  \n",
    "  \n",
    "  assign(paste0(\"spe_\", sample_name), spe)\n",
    "  #print progress check\n",
    "  cat(\"Loaded:\", sample_name, \"\\n\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#1.6 load arbitrary batch \n",
    "\n",
    "load_all_data <- \"/mnt/scratchc/fmlab/lythgo02/visium_data/arbitrary/nnsvg\"\n",
    "# List files in the directory\n",
    "files <- list.files(path = load_all_data, pattern = \"^nnsvg\", full.names = TRUE)\n",
    "\n",
    "# Print out the list of files\n",
    "print(files)\n",
    " \n",
    "#assign names so they are loaded into the spe with the correct identifiers\n",
    "arbitrary <- c(\"SITSA1\",\"SITSA2\",\"SITSA4\", \"SITSB1\", \"SITSB3\", \n",
    "                 \"SITSC2\",\"SITSC4\",\"SITSD1\",\"SITSD2\", \"SITSD4\",\n",
    "                 \"SITSE2\",\"SITSG2\", \"SITSH2\")\n",
    "\n",
    "names(files) <- arbitrary \n",
    "\n",
    "# Loop through each file and process it\n",
    "for (sample_name in names(files)) {\n",
    " # Read the Visium data\n",
    "  spe <- readRDS(files[sample_name])\n",
    "    # Add unique barcodes\n",
    "  spe$barcode <- rownames(colData(spe))\n",
    "  spe$barcodeid <- gsub(\"-1$\", paste0(\"-\", sample_name), spe$barcode)\n",
    "  rownames(colData(spe)) <- spe$barcodeid\n",
    "  spe$sample_id <- sample_name\n",
    "  \n",
    "  # Store the processed sample in the list\n",
    "  spe_list[[sample_name]] <- spe \n",
    "  \n",
    "  \n",
    "  assign(paste0(\"spe_\", sample_name), spe)\n",
    "  #print progress check\n",
    "  cat(\"Loaded:\", sample_name, \"\\n\")\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#1.6 bind all into one spatial object \n",
    "\n",
    "\n",
    "    spe <- do.call(cbind, spe_list) #wont work because different row numbers # wont work if row numbers differ"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. To process all samples together in a slurm submission (either up to nnsvg or inc nnsvg)\n",
    "#make sure you are updating and using the scratch scripts, there are also copies on mnt but outdated\n",
    "\n",
    "1_sample_specific_filtering_wholeworkflow.r isn't actually the whole workflow as the end is commented out so that next stages are run via 2_end....sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#2.0\n",
    "\n",
    "sbatch /mnt/scratchc/fmlab/lythgo02/visium_data/scripts/submit_spatial_scripts.sh\n",
    "\n",
    "#which is \n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "#SBATCH --error=nnSVG.err\n",
    "#SBATCH --time=0200:00:00             \n",
    "#SBATCH --cpus-per-task=32      \n",
    "#SBATCH --mem=120G                   \n",
    "#SBATCH --partition=epyc     \n",
    "\n",
    "\n",
    "# Run the R script\n",
    "#Rscript /mnt/scratchc/fmlab/lythgo02/visium_data/scripts/1_sample_specific_filtering_wholeworkflow.r "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#there is also a script called 2_end_stage_hvg_nnsvg_singlesamplejob.r which is the end section of the workflow to submit single jobs for each sample for hvg and nnsvg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1_sample_specific_filtering_wholeworkflow.r runs the following steps:\n",
    "Workflow Summary:\n",
    "\n",
    "\n",
    " - Load Visium spatial transcriptomics data per sample.\n",
    " - Add unique barcodes to each spot for tracking\n",
    " - Combine all samples into one SpatialExperiment (SPE) object.\n",
    " - Gene Annotationvia Ensembl\n",
    " - Filter out genes with zero counts.\n",
    " - Spot Filtering and QC:\n",
    "    -  Keep only spots overlaying tissue.\n",
    " - Calculate per-spot QC metrics:\n",
    "     - Number of detected genes,\n",
    "     - Total UMIs (library size),\n",
    "     - Percentage of mitochondrial UMIs.\n",
    " - Generate QC summary tables and save intermediate results.\n",
    " - Split cohort into log and arbitrary:\n",
    "     - A few of he samples have high mito content so arbitrary threshold across all doesn't work, if we expect some mitochondrial content due to systematic/technical processes then we would expect this to show up fairly uniformly across samples. For these samples I apply a sample specific threshold. Exclude spots that are more than a particular standard deviation away from the mean. \n",
    "     - Mitochondrial % filter:\n",
    "        -  In the “log” cohort, uses a sample-specific, robust outlier detection method on the log-transformed mitochondrial percentages (median ± MAD).\n",
    "        -  In the “arbitrary” cohort, applies a fixed, arbitrary threshold (e.g., >20%).\n",
    "    -  Library size and gene detection filters: use fixed thresholds for all samples regardless of cohort.\n",
    "         - Library Size QC: Apply a threshold (e.g., <300 UMIs) to flag low-quality spots.\n",
    "          -  Gene Detection QC: threshold (e.g., <300 detected genes) to flag poor spots.\n",
    " - Save QC-filtered SPE objects and plots at:\n",
    "  - visium_data/single_sample_from_filtering/20250321_spe_sub_sampleID.rds\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# 2.1 then process each sample separately for hvg and nnsvg via slurm job using following script (has the following steps in 2_end_stage_hvg_nnsvg_singlesamplejob.r)\n",
    "\n",
    "# note after separating out individual samples, SITSA1 and SITSD3 have rows with 0 count (causing nnsvg to fail) for:\n",
    "#SITSA1 - HNRNPCL3, TMOD4, REG1A USP17L10, PVRIG CNTNAP3B, ZP1, C13orf46 ZP2, PAGE2, TGIF2LY, PCDH11Y \n",
    "#SITSD3 - HNRNPCL3, PRAMEF33, UBE2U, CA14 AL590560.2, WDR64, SPDYA, NMS, ZDHHC19, USP17L10, HTN3, PCBD2, HIST1H4E, PVRIG \n",
    "         #ASZ1, AKR1C4, OOSP1, ZP1, DEFB131B, C13orf46, GOLGA6L6, TEPP, ALOX15, ANKRD30B, ZNF541, BPIFB2, BPIFB6, BPIFB3 \n",
    "         #USP41, CRYBB2, GAGE1, PAGE2, SLC25A5, TGIF2LY, PRKY \n",
    "#so the following is run again within the script\n",
    "\n",
    "# remove zero expression genes\n",
    "ix_zero_genes <- rowSums(counts(spe_SITSD3)) == 0\n",
    "table(ix_zero_genes)\n",
    "\n",
    "# identify genes with 0 expression\n",
    "true <- ix_zero_genes[ix_zero_genes]\n",
    "\n",
    "if (sum(ix_zero_genes) > 0) {\n",
    "  spe_SITSD3 <- spe_SITSD3[!ix_zero_genes, ]}\n",
    "\n",
    "# remove spots with zero expression\n",
    "ix_zero_spots <- colSums(counts(spe_SITSA1)) == 0\n",
    "table(ix_zero_spots) #0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#2.1 \n",
    "#make sure you change which line is commented out in submit_spatial_scripts.sh so 2_end_stage_hvg_nnsvg_singlesamplejob.r is submitted\n",
    "\n",
    "for sample in SITSA1 SITSA2 SITSA3 SITSA4 SITSB1 SITSB2 SITSB3 SITSB4 SITSC1 SITSC2 SITSC3 SITSC4 SITSD1 SITSD2 SITSD3 SITSD4 SITSE2 SITSE4 SITSF2 SITSF4 SITSG2 SITSH2; \n",
    "do     sbatch ../scripts/submit_spatial_scripts.sh $sample; done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2_end_stage_hvg_nnsvg_singlesamplejob.r:  \n",
    " - Loads the sample-specific RDS file.  \n",
    " - Removes genes and spots with zero expression.  \n",
    " - Applies a gene filter function (filter_genes) without mitochondrial filtering.  \n",
    " - Normalizes counts and computes library size factors.  \n",
    " - Identifies highly variable genes (HVGs) using modelGeneVar and flags them. Saves intermediate data objects.  \n",
    " - Checks for normalized log-counts, then runs the nnSVG spatial gene variability method.  \n",
    " - Saves the final object with nnSVG results."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2.3 then open interactive R session and proceed as follows\n",
    "We will take the top ranking HVGs and top ranking SVGs, reduce the dimensions with PCA on each, and then integrate and cluster\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "srun --pty --job-name=interactive_r --mem=68G --cpus-per-task=4 --partition=epyc --time=100:00:00 bash\n",
    "\n",
    "micromamba activate r_spatial\n",
    "\n",
    "R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#2.4 fitting top HVGs\n",
    "\n",
    "#knitr::opts_chunk$set(echo = TRUE)\n",
    "library(spatialLIBD)\n",
    "library(nnSVG)\n",
    "library(SpatialExperiment)  \n",
    "library(scater)\n",
    "library(AnnotationHub)\n",
    "library(tidyverse)\n",
    "library(ggspavis)\n",
    "library(scran)\n",
    "library(DT)\n",
    "library(HDF5Array)\n",
    "projDir = \"/mnt/scratchc/fmlab/lythgo02/visium_data/\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load Pre-processed Samples:\n",
    "List files in a directory matching \"spe_sub_modelgenevar\" (these contain samples processed for highly variable genes, HVGs).\n",
    "Loop through the files, load each .rds file into R, and store the loaded data into spe_list (a list of SpatialExperiment objects).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# ====== Load HVG-processed samples ======\n",
    "load_all_data <- \"/mnt/scratchc/fmlab/lythgo02/visium_data/single_sample_from_filtering/\" \n",
    "# Alternative path (uncomment if needed)\n",
    "# load_all_data <- \"/mnt/nas-data/fmlab/group_folders/lythgo02/visium_data/single_sample_from_filtering/\" \n",
    "\n",
    "# Sample IDs\n",
    "samplesheet <- c(\"SITSA1\", \"SITSA2\", \"SITSA3\", \"SITSA4\",\n",
    "                 \"SITSB1\", \"SITSB2\", \"SITSB3\", \"SITSB4\", \n",
    "                 \"SITSC1\", \"SITSC2\", \"SITSC3\", \"SITSC4\", \n",
    "                 \"SITSD1\", \"SITSD2\", \"SITSD3\", \"SITSD4\",\n",
    "                 \"SITSE2\", \"SITSE4\", \"SITSF2\", \"SITSF4\",\n",
    "                 \"SITSG2\", \"SITSH2\")  \n",
    "\n",
    "# Load HVG-filtered RDS files\n",
    "hvg_files <- list.files(path = load_all_data, pattern = \"spe_sub_modelgenevar\", full.names = TRUE)\n",
    "names(hvg_files) <- samplesheet\n",
    "\n",
    "# Load HVG objects into a list\n",
    "spe_list <- list()\n",
    "for (sample_name in names(hvg_files)) {\n",
    "  spe <- readRDS(hvg_files[[sample_name]])\n",
    "  spe_list[[sample_name]] <- spe\n",
    "  cat(\"Loaded HVG sample:\", sample_name, \"\\n\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dimensionality reduction on top HVGs \n",
    "Top HVGs have been determined using getTopHVGs() from scran which returns top 10% (prop=0.1) of genes with highest biological variance after discarding those with bio <= 0\n",
    "Extract HVGs and Run PCA & UMAP (on PCAs) and save back to list\n",
    "\n",
    "Use PCA to reduce the dimensions of our dataset to assist clustering and UMAP to further reduce the principal components (PCs) in a two-dimensional space and produce better visualisations for the PCA.\n",
    "runPCA() function runs PCA on a SCE object, and returns an updated version of the single cell object with the PCA result added to the reducedDim slot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "# ====== Extract HVGs and run PCA/UMAP ======\n",
    "set.seed(123)\n",
    "\n",
    "# Store HVG gene symbols\n",
    "hvg <- list()\n",
    "for (sample_name in names(spe_list)) {\n",
    "  hvg[[sample_name]] <- rowData(spe_list[[sample_name]]) %>%\n",
    "    as.data.frame() %>%\n",
    "    filter(hvg == TRUE) %>%\n",
    "    pull(symbol)\n",
    "\n",
    "  # PCA\n",
    "  spe_list[[sample_name]] <- runPCA(spe_list[[sample_name]], subset_row = hvg[[sample_name]], name = \"PCA_hvg\")\n",
    "}\n",
    "\n",
    "# Run UMAP on top PCs\n",
    "set.seed(987)\n",
    "for (sample_name in names(spe_list)) {\n",
    "  spe_list[[sample_name]] <- runUMAP(spe_list[[sample_name]], dimred = \"PCA_hvg\", name = \"UMAP_hvg\")\n",
    "  colnames(reducedDim(spe_list[[sample_name]], \"UMAP_hvg\")) <- c(\"UMAP_hvg1\", \"UMAP_hvg2\")\n",
    "}\n",
    "\n",
    "# (Optional) Plot PCA\n",
    "plots <- list()\n",
    "for (sample_name in names(spe_list)) {\n",
    "  sample_data <- spe_list[[sample_name]]\n",
    "  pca_df <- as.data.frame(reducedDim(sample_data, \"PCA_hvg\"))\n",
    "  pca_df$layer <- colData(sample_data)$ground_truth\n",
    "\n",
    "  plots[[sample_name]] <- ggplot(pca_df, aes(x = PC1, y = PC2, colour = layer)) +\n",
    "    geom_point(size = 0.5) +\n",
    "    scale_colour_brewer(type = \"qual\") +\n",
    "    labs(title = paste(\"PCA (Sample:\", sample_name, \")\"),\n",
    "         x = \"PC1\", y = \"PC2\", colour = \"Layers\") +\n",
    "    theme_classic()\n",
    "}\n",
    "\n",
    "# Define the directory where you want to save the plots\n",
    "#save_dir <- \"/mnt/scratchc/fmlab/lythgo02/visium_data/single_sample_from_filtering/\"\n",
    "\n",
    "#for (i in seq_along(plots)) {\n",
    "  # Define the file name based on the sample_id for uniqueness, with the full path\n",
    " # file_name <- paste0(save_dir, \"plots/plot_hvg_pca\", spe_list[[i]]$sample_id[1], \".png\")\n",
    "  \n",
    "  # Save the plot using ggsave\n",
    "  #ggsave(filename = file_name, plot = plots[[i]], width = 8, height = 6)\n",
    "#}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# or for a single sample at a time\n",
    "\n",
    "# 1. Extract HVGs\n",
    "hvg <- as.data.frame(rowData(spe_SITSA1)) %>%\n",
    "  filter(hvg == TRUE) %>%\n",
    "  pull(symbol)\n",
    "\n",
    "# 2. Run PCA on HVGs\n",
    "spe_SITSA1 <- runPCA(spe_SITSA1, subset_row = hvg, name=\"PCA_hvg\")\n",
    "\n",
    "# 3. Run UMAP on top PCs\n",
    "set.seed(987)\n",
    "spe_SITSA1 <- runUMAP(spe_SITSA1, dimred = \"PCA_hvg\", name=\"UMAP_hvg\")\n",
    "\n",
    "# 4. Check structure\n",
    "print(reducedDimNames(spe_SITSA1))\n",
    "print(dim(reducedDim(spe_SITSA1, \"UMAP_hvg\")))\n",
    "\n",
    "# 5. Rename columns for plotting\n",
    "colnames(reducedDim(spe_SITSA1, \"UMAP_hvg\")) <- paste0(\"UMAP_hvg\", 1:2)\n",
    "\n",
    "# 6. Plot PCA\n",
    "ggplot(data = as.data.frame(reducedDim(spe_SITSA1, \"PCA_hvg\")),\n",
    "       aes(x = PC1, y = PC2, colour = colData(spe_SITSA1)$ground_truth)) +\n",
    "  geom_point(size = 0.5) +\n",
    "  scale_colour_brewer(type = \"qual\") +\n",
    "  labs(title = paste(\"Reduced dimensions: PCA (Sample:\", spe_SITSA1$sample_id[1], \")\"),\n",
    "       x = \"PC1\", y = \"PC2\", colour = \"Layers\") +\n",
    "  theme_classic()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2.6 Get top SVGs\n",
    "\n",
    "nnsvg\n",
    "The results are stored in the rowData of the SpatialExperiment object.\n",
    "The main results of interest are:\n",
    "LR_stat: likelihood ratio (LR) statistics used to rank SVGs\n",
    "rank: rank of top SVGs according to LR statistics\n",
    "pval: approximate p-values\n",
    "padj: approximate p-values adjusted for multiple testing\n",
    "prop_sv: effect size defined as proportion of spatial variance\n",
    "\n",
    "First load each spe object after having run nnsvg on the cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# ====== Load nnSVG-ranked samples ======\n",
    "svg_files <- list.files(path = load_all_data, pattern = \"spe_sub_nnSVG\", full.names = TRUE)\n",
    "names(svg_files) <- samplesheet\n",
    "\n",
    "spe_list_nnsvg <- list()\n",
    "for (sample_name in names(svg_files)) {\n",
    "  spe_svg <- readRDS(svg_files[[sample_name]])\n",
    "  spe_list_nnsvg[[sample_name]] <- spe_svg\n",
    "  cat(\"Loaded nnSVG sample:\", sample_name, \"\\n\")\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "get top SVGs for each sample using user defined function"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run PCA and UMAP on top SVGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "# ====== Compute PCA/UMAP on SVGs and find common genes ======\n",
    "set.seed(456)\n",
    "\n",
    "svg <- list()\n",
    "common_genes_list <- list()\n",
    "\n",
    "for (sample_name in names(spe_list_nnsvg)) {\n",
    "  spe_svg <- spe_list_nnsvg[[sample_name]]\n",
    "  #spe_hvg <- spe_list[[sample_name]]\n",
    "\n",
    "  # Add SVG flag if not already present\n",
    "  rowData(spe_svg)$svg <- rowData(spe_svg)$rank <= 1000\n",
    "\n",
    "  # Extract SVGs (gene IDs)\n",
    "  svg[[sample_name]] <- rownames(spe_svg)[rowData(spe_svg)$svg == TRUE]\n",
    "\n",
    "  # PCA and UMAP on SVGs\n",
    "  spe_svg <- runPCA(spe_svg, subset_row = svg[[sample_name]], name = \"PCA_svg\")\n",
    "  spe_svg <- runUMAP(spe_svg, dimred = \"PCA_svg\", name = \"UMAP_svg\")\n",
    "  colnames(reducedDim(spe_svg, \"UMAP_svg\")) <- c(\"UMAP_svg1\", \"UMAP_svg2\")\n",
    "\n",
    "  # Update object\n",
    "  spe_list_nnsvg[[sample_name]] <- spe_svg\n",
    "\n",
    "  # Find overlap with HVGs\n",
    "  #common_genes_list[[sample_name]] <- intersect(hvg[[sample_name]], svg[[sample_name]])\n",
    "}\n",
    "\n",
    "#metadata table\n",
    "group_df <- data.frame(\n",
    "  sample_id = c(\n",
    "    \"SITSC1\", \"SITSD2\", \"SITSH2\", \"SITSC2\", \"SITSA3\",\n",
    "    \"SITSA1\", \"SITSD4\", \"SITSG2\", \"SITSB3\",\n",
    "    \"SITSB1\", \"SITSD1\", \"SITSA4\", \"SITSB4\", \"SITSF2\",\n",
    "    \"SITSC3\", \"SITSF4\", \"SITSE4\",\n",
    "    \"SITSB2\",\n",
    "    \"SITSD3\",\n",
    "    \"SITSC4\", \"SITSA2\", \"SITSE2\"\n",
    "  ),\n",
    "    subgroup = c(\n",
    "    rep(\"gBRCA1m-ov\", 5),\n",
    "    rep(\"gBRCA2m-ov\", 4),\n",
    "    rep(\"HRP-ov\", 5),\n",
    "    rep(\"non-BRCA-HRD-ov\", 3),\n",
    "    \"gBRCA2m-om\",\n",
    "    \"gBRCA1m-om\",\n",
    "    rep(\"HRP-om\", 3)\n",
    "  )\n",
    ")\n",
    "\n",
    "gene_summary <- data.frame(\n",
    "  sample_id=group_df$sample_id,\n",
    "  subgroup=group_df$subgroup,\n",
    "  hvg_n=sapply(group_df$sample_id, function(sample)\n",
    "  length(hvg[[sample]])),\n",
    "  hvg_in_svg_n=sapply(group_df$sample, function(sample)\n",
    "  length(common_genes_list[[sample]])))\n",
    "\n",
    "gene_summary$genotype <- sub(\"-o.*\", \"\", gene_summary$subgroup)\n",
    "\n",
    "gene_summary$proportion <- gene_summary$hvg_in_svg_n / gene_summary$hvg_n\n",
    "\n",
    "\n",
    "\n",
    "# Set output directory\n",
    "save_dir <- \"/home/lythgo02/Spatial/plots/\"\n",
    "\n",
    "\n",
    "# 1️⃣ Boxplot\n",
    "p1 <- ggplot(gene_summary, aes(x = genotype, y = proportion, fill = genotype)) +\n",
    "  geom_boxplot(outlier.shape = NA, alpha = 0.5) +\n",
    "  geom_jitter(width = 0.2, size = 2) +\n",
    "  labs(\n",
    "    title = \"Proportion of HVGs also found in SVGs\",\n",
    "    x = \"Genotype\",\n",
    "    y = \"Proportion HVGs in SVGs\"\n",
    "  ) +\n",
    "  theme_minimal(base_size = 14) +\n",
    "  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +\n",
    "  scale_fill_brewer(palette = \"Set2\")\n",
    "\n",
    "# Save boxplot\n",
    "ggsave(filename = file.path(save_dir, \"hvg_svg_proportion_boxplot.png\"), plot = p1, width = 8, height = 6)\n",
    "\n",
    "# 2️⃣ Per-sample line plot\n",
    "p2 <- ggplot(gene_summary, aes(x = reorder(sample_id, proportion), y = proportion, color = genotype)) +\n",
    "  geom_point(size = 3) +\n",
    "  geom_segment(aes(xend = reorder(sample_id, proportion), y = 0, yend = proportion), alpha = 0.4) +\n",
    "  labs(\n",
    "    title = \"Per-sample HVG-SVG Proportion\",\n",
    "    x = \"Sample ID\",\n",
    "    y = \"Proportion\"\n",
    "  ) +\n",
    "  theme_minimal(base_size = 13) +\n",
    "  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +\n",
    "  scale_color_brewer(palette = \"Set2\")\n",
    "\n",
    "# Save per-sample plot\n",
    "ggsave(filename = file.path(save_dir, \"hvg_svg_proportion_by_sample.png\"), plot = p2, width = 10, height = 6)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "\n",
    "HVGs = Highly Variable Genes: These are genes showing high variation across spots/samples.\n",
    "SVGs = Spatially Variable Genes: Genes whose expression varies with spatial location (spots).\n",
    "Overlap (hvg_in_svg_n): Number of HVGs that are also spatially variable in that sample.\n",
    "\n",
    "Biological / Analytical insights:\n",
    "High Proportion (~0.7-1.0): Most HVGs in these samples are spatially variable, implying that gene variability is largely driven by spatial differences in tissue.\n",
    "Example: SITSA1 has a perfect 1.0 proportion — all its HVGs are spatially variable.\n",
    "This suggests spatial patterns strongly influence gene variability here.\n",
    "Lower Proportion (~0.45-0.6): A lower fraction of HVGs are spatially variable.\n",
    "Example: SITSD4 with 0.45 proportion means more than half HVGs are variable but not necessarily spatially patterned — possibly due to cell type variation or technical noise.\n",
    "Inter-sample variability: The proportion varies between samples, which might reflect biological differences (e.g., different tissue regions, pathological states) or technical differences (e.g., sequencing depth).\n",
    "\n",
    "SVGs capture genes whose expression varies spatially, which can reveal tissue structure and microenvironment = primary focus so focussing on these (esp given that for most samples >70% HVGs are in SVGs)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Integrate SVGs and HVGs (or just take the SVGs if the HVGs are in SVGs for each sample)........I looked into the following statement/paper and it doesn't actually seem to comment on SVG/HVG integration\n",
    "A recent benchmark paper (Li et al. 2022) showed that integrating HVGs and SVGs to generate a combined set of features can improve downstream clustering performance in STx data. This confirms that SVGs contain additional biologically relevant information that is not captured by HVGs in these datasets. For example, a simple way to combine these features is to concatenate columns of principal components (PCs) calculated on the set of HVGs and the set of SVGs (excluding overlapping HVGs), and then using the combined set of features for further downstream analyses (Li et al. 2022)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cluster gene expression via scRNAseq methods\n",
    "Graph-based clustering using the Walktrap method implemented in scran (Lun, McCarthy, and Marioni 2016), builds shared nearest neighbour (SNN) using top 50 PCs calculated on the set of top SVGs.Assume that biologically informative spatial distribution patterns of cell types can be detected from the molecular features (gene expression).\n",
    "\n",
    "NEED TO CHECK IF I AM USING TOP 50 PCS OR ALL PC ETC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# graph-based clustering to identify spatial domains \n",
    "set.seed(123)\n",
    "spot_counts <- sapply(spe_list_nnsvg, ncol)\n",
    "spot_counts_df <- data.frame(spot_counts, \n",
    "                             guide_k = ifelse(spot_counts>2000,15,10)) # if spot number >2000 k = 15, otherwise k=10\n",
    "#set numbers of nearest neighbours for clustering, k=10 common, increase to prevent overfitting and large cluster numbers/too few spots in\n",
    "#15 avoids clusters with fewer than 30 spots, set k=10 for samples with fewer spots to avoid oversmoothing\n",
    "cluster_summary <- list()\n",
    "\n",
    "for (sample in names(spe_list_nnsvg)) {\n",
    "  k_value=spot_counts_df[sample, \"guide_k\"]\n",
    "  g <- buildSNNGraph(spe_list_nnsvg[[sample]], k = k_value, use.dimred = \"PCA_svg\")  #builds graph using PCAs from nnSVGs\n",
    "  g_walk <- igraph::cluster_walktrap(g) #applies clustering algorithm, default number of steps = 4, increase to prevent overfitting and clusters with too few spots in\n",
    "  clus <- g_walk$membership  # extracts cluster assignments as vector for each spot\n",
    "  cluster_summary[[sample]] <- table(clus)  # store cluster sizes per sample in a table in cluster_summary\n",
    " colLabels(spe_list_nnsvg[[sample]]) <- factor(clus) #store labels\n",
    "}\n",
    "\n",
    "#for each sample, count how many unique clusters were found\n",
    "cluster_counts <- data.frame(\n",
    "  sample_id = names(spe_list_nnsvg),\n",
    "  n_clusters = sapply(spe_list_nnsvg, function(s) length(unique(colLabels(s))))\n",
    ")\n",
    "\n",
    "#add meta data\n",
    "cluster_counts$subgroup <- group_df$subgroup[match(cluster_counts$sample_id, group_df$sample_id)]\n",
    "cluster_counts$genotype <- sub(\"-o.*\", \"\", cluster_counts$subgroup)\n",
    "\n",
    "\n",
    "library(ggplot2)\n",
    "#plot cluster number per sample, coloured by genotype\n",
    "p3 <-ggplot(cluster_counts, aes(x = reorder(sample_id, n_clusters), y = n_clusters, color = genotype)) +\n",
    "  geom_point(size = 3) +\n",
    "  geom_segment(aes(xend = reorder(sample_id, n_clusters), y = 0, yend = n_clusters), alpha = 0.4) +\n",
    "  labs(\n",
    "    title = \"Per-sample Cluster Diversity\",\n",
    "    x = \"Sample ID\",\n",
    "    y = \"Number of Clusters\"\n",
    "  ) +\n",
    "  theme_minimal(base_size = 13) +\n",
    "  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +\n",
    "  scale_color_brewer(palette = \"Set2\")\n",
    "\n",
    "ggsave(filename = file.path(save_dir, \"spatial_clusters.png\"), plot = p3, width = 10, height = 6)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot spatial domains\n",
    "Plot PCA coloured by spatial domain label\n",
    "Plot UMAP coloured by spatial domain label "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "for (sample in names(spe_list_nnsvg)) {\n",
    "  spe <- spe_list_nnsvg[[sample]]\n",
    "  plotSpots(spe, annotate = \"label\")\n",
    "  ggsave(filename =  file.path(save_dir, paste0(sample, \"_SVGs_clustered_walktrap.png\")), width = 6, height = 6)\n",
    "  plotDimRed(spe, plot_type=\"UMAP_svg\", annotate=\"label\")\n",
    "  ggsave(filename =  file.path(save_dir, paste0(sample, \"_umap.png\")), width = 6, height = 6)\n",
    "  plotDimRed(spe, plot_type=\"PCA_svg\", annotate=\"label\")\n",
    "  ggsave(filename =  file.path(save_dir, paste0(sample, \"_pca.png\")), width = 6, height = 6)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "The approach described in the tutorial then suggests adding the raw (or normalized) x-y spatial coordinates as additional dimensions during clustering — concatenated with your molecular PCs (like the PCA from SVGs or HVGs). \n",
    "This directly and explicitly forces the clustering to also consider physical proximity in space.\n",
    "\n",
    "The following section combines PCAs and xy-coordinates as a slot in the spatial experiment named \"PCA_svg_plus_xy\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "#For example:\n",
    "\n",
    "# Combine PCA of SVGs + spatial coordinates\n",
    "spot_counts <- sapply(spe_list_nnsvg, ncol)\n",
    "spot_counts_df <- data.frame(spot_counts, \n",
    "                             guide_k = ifelse(spot_counts>2000,15,10)) # if spot number >2000 k = 15, otherwise k=10\n",
    "\n",
    "\n",
    "for (sample in names(spe_list_nnsvg)) {\n",
    "  spe <- spe_list_nnsvg[[sample]]  # extract sample object\n",
    "  # Extract PCA\n",
    "  pcs <- reducedDim(spe, \"PCA_svg\")\n",
    "  # Extract and relabel coords to match colnames (with sample ID)\n",
    "  coords <- spatialCoords(spe)\n",
    "  rownames(coords) <- paste0(sub(\"-1$\", \"\", rownames(coords)), \"-\", sample)\n",
    "  # Sanity check: ensure rownames match\n",
    "  stopifnot(identical(rownames(pcs), rownames(coords)))\n",
    "  # Combine PCA + spatial coordinates\n",
    "  combined <- cbind(pcs, coords)\n",
    "  reducedDim(spe, \"PCA_svg_plus_xy\") <- combined\n",
    "  # Set k from spot_counts_df (ensure rownames are correct)\n",
    "  k_value <- spot_counts_df[sample, \"guide_k\"]\n",
    "  # Build SNN graph using reducedDim\n",
    "  g <- buildSNNGraph(spe, k = k_value, use.dimred = \"PCA_svg_plus_xy\")\n",
    "  # Apply Walktrap clustering\n",
    "  g_walk <- igraph::cluster_walktrap(g)\n",
    "  clus <- g_walk$membership\n",
    "  # Assign clusters\n",
    "  colLabels(spe) <- factor(clus)\n",
    "  # Save updated object back into list\n",
    "  spe_list_nnsvg[[sample]] <- spe\n",
    "}\n",
    "\n",
    "\n",
    "#for each sample, count how many unique clusters were found\n",
    "cluster_counts <- data.frame(\n",
    "  sample_id = names(spe_list_nnsvg),\n",
    "  n_clusters = sapply(spe_list_nnsvg, function(s) length(unique(colLabels(s))))\n",
    ")\n",
    "\n",
    "#add meta data\n",
    "cluster_counts$subgroup <- group_df$subgroup[match(cluster_counts$sample_id, group_df$sample_id)]\n",
    "cluster_counts$genotype <- sub(\"-o.*\", \"\", cluster_counts$subgroup)\n",
    "\n",
    "\n",
    "library(ggplot2)\n",
    "#plot cluster number per sample, coloured by genotype\n",
    "p3 <-ggplot(cluster_counts, aes(x = reorder(sample_id, n_clusters), y = n_clusters, color = genotype)) +\n",
    "  geom_point(size = 3) +\n",
    "  geom_segment(aes(xend = reorder(sample_id, n_clusters), y = 0, yend = n_clusters), alpha = 0.4) +\n",
    "  labs(\n",
    "    title = \"Per-sample Cluster Diversity\",\n",
    "    x = \"Sample ID\",\n",
    "    y = \"Number of Clusters\"\n",
    "  ) +\n",
    "  theme_minimal(base_size = 13) +\n",
    "  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +\n",
    "  scale_color_brewer(palette = \"Set2\")\n",
    "\n",
    "ggsave(filename = file.path(save_dir, \"spatial_clusters.png\"), plot = p3, width = 10, height = 6)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save results efficiently in HDF5 format, useful for large datasets as it is scalable, avoiding memory over-load. \n",
    "Reload the objects with loadHDF5SummarizedExperiment()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(HDF5Array)\n",
    "#saved locally on workstation\n",
    "dir.create(\"Home/Documents/saved_spe_h5\", showWarnings = FALSE)\n",
    "\n",
    "for (sample in names(spe_list_nnsvg[13:22])) {\n",
    "  spe <- spe_list_nnsvg[[sample]]\n",
    "  saveHDF5SummarizedExperiment(\n",
    "    spe,\n",
    "    dir = file.path(\"saved_spe_h5\", paste0(sample, \"_spe_nnsvg_clusterDE\")),\n",
    "    replace = TRUE\n",
    "  )\n",
    "}\n",
    "\n",
    "\n",
    "# Load HVG-filtered RDS files\n",
    "cluster_files <- list.files(path = \"/home/lythgo02/Documents/saved_spe_h5/\", pattern = \"_spe_nnsvg_clusterDE\", full.names = TRUE)\n",
    "spe_list <- list()\n",
    "names(cluster_files) <- samplesheet\n",
    "\n",
    "for (sample in names(cluster_files)){\n",
    "  spe <- loadHDF5SummarizedExperiment(cluster_files[[sample]])\n",
    "  spe_list[[sample]] <- spe\n",
    "}\n",
    "\n",
    "\n",
    "library(zellkonverter)\n",
    "\n",
    "# Assuming your object is called `spe`\n",
    "\n",
    "for (sample in names(spe_list_nnsvg[1:22])) {\n",
    "  spe <- spe_list_nnsvg[[sample]]\n",
    "  zellkonverter::writeH5AD(\n",
    "    spe,\n",
    "    file = file.path(\"/mnt/scratchc/fmlab/lythgo02/Spatial/saved_spe_h5ad\", paste0(sample, \"_spe_nnsvg_clusterDE.h5ad\"))\n",
    "  )\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "NEED TO DECIDE WHETHER TO DO SPATIALLY INFORMED (EXPLICIT) OR NOT \n",
    "get more spatially compact/defined clusters with the spatial coordinates, how to decide, try comparing to the HnE images and see if either matches better \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next steps\n",
    "Differential Expression Between Clusters\n",
    "Gene Ontology / Pathway Enrichment\n",
    "Use cluster locations (from colData/colLabels) to compare with image-derived features:\n",
    "    In H&E: align clusters with morphology (e.g., necrosis, ductal regions, stroma).\n",
    "    In mIF: are immune-rich clusters spatially aligned with CD3+, CD8+, or macrophage-rich zones?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Perform gene ontology enrichment abnalysis on cluster-specific DGEs:\n",
    "For each sample, and each cluster within the sample, perform:\n",
    "    Differential expression (cluster vs all others)\n",
    "    Select DEGs (FDR < 0.05)\n",
    "    Map gene symbols → Entrez IDs\n",
    "    Run GO enrichment analysis (Biological Process category)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(clusterProfiler)\n",
    "library(org.Hs.eg.db)\n",
    "library(scater)\n",
    "\n",
    "\n",
    "go_results_list <- list()\n",
    "\n",
    "for (sample in names(spe_list_nnsvg)) {\n",
    "  spe <- spe_list_nnsvg[[sample]]\n",
    "  clusters <- colLabels(spe)\n",
    "\n",
    "  # DE between clusters\n",
    "  de <- findMarkers(spe, groups = clusters, test.type = \"t\")\n",
    "\n",
    "  # Loop through each cluster in that sample\n",
    "  sample_go_results <- list()\n",
    "\n",
    "  for (cluster_id in names(de)) {\n",
    "    # Get DEGs for this cluster vs. others\n",
    "    top_genes <- rownames(de[[cluster_id]])[de[[cluster_id]]$FDR < 0.05]\n",
    "\n",
    "    # Map to Entrez IDs\n",
    "    gene_df <- bitr(top_genes, fromType = \"SYMBOL\", toType = \"ENTREZID\", OrgDb = org.Hs.eg.db)\n",
    "\n",
    "    # Skip if no genes mapped\n",
    "    if (nrow(gene_df) < 5) next\n",
    "\n",
    "    # GO enrichment (Biological Process)\n",
    "    ego <- enrichGO(\n",
    "      gene = gene_df$ENTREZID,\n",
    "      OrgDb = org.Hs.eg.db,\n",
    "      ont = \"BP\",\n",
    "      pAdjustMethod = \"BH\",\n",
    "      qvalueCutoff = 0.05,\n",
    "      readable = TRUE\n",
    "    )\n",
    "\n",
    "    sample_go_results[[cluster_id]] <- ego\n",
    "  }\n",
    "\n",
    "  go_results_list[[sample]] <- sample_go_results\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#NOTE SAVED IN GO RESULTS IN FOLDER "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "Run clustering and GO enrichment within each sample (as you already do).\n",
    "\n",
    "For each subgroup, summarize:\n",
    "    Number of clusters per sample (diversity)\n",
    "    Recurrent GO terms across clusters and samples\n",
    "    Possibly score clusters by significance or similarity\n",
    "    Identify patterns across samples in the same subgroup:\n",
    "    Shared GO terms\n",
    "    Shared cluster structures (e.g., similar cell states)\n",
    "    Variability within/between subgroups\n",
    "\n",
    "Advantages:\n",
    "Preserves spatial resolution and biological individuality of each sample.Handles sample heterogeneity well. Identifies dentify subgroup-specific recurrent patterns, not just averages"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
